### PR TITLE
Set order on Landing page save

### PR DESCRIPTION
**Pull request by amitaibu**
Users can now drag and drop the Paragraphs, and on save weight should be re-saved if it has changed.

image

### CODE REVIEW

File: Web/Controller/LandingPages.hs [hunk 1]
```diff
@@ -38,6 +38,34 @@ instance Controller LandingPagesController where
...
+                        uuids -> do
+                            -- We need to update the weight of the paragraphs,
+                            -- So load them.
+                            landingPage <- fetchLandingPageWithParagraphs landingPageId
+
+                            -- Iterate over all paragraphs, and update the weight.
+                            -- @todo: Iterate over all paragraphs.
+                            forEach landingPage.paragraphCta \paragraph -> do
+                                let uuid = unpackId paragraph.id
+                                let weight = elemIndex uuid uuids |> fromMaybe 0
```
Comment (amitaibu) [inline Web/Controller/LandingPages.hs:59]: image

But even if I remove this line, I'll get another error on the `updateRecord` below

image

Comment (amitaibu) [reply Web/Controller/LandingPages.hs:59]: Typo in the variable name :/

File: Web/Controller/LandingPages.hs [hunk 2]
```diff
@@ -38,6 +38,38 @@ instance Controller LandingPagesController where
...
+                        uuids -> do
+                            -- We need to update the weight of the paragraphs,
+                            -- So load them.
+                            landingPage <- fetchLandingPageWithParagraphs landingPageId
+
+                            -- Iterate over all paragraphs, and update the weight.
+                            forEach landingPage.paragraphCtas updateParagraph
+
+                            -- @todo: If I uncomment this, it results with a compiler error.
+                            -- forEach landingPage.paragraphQuotes updateParagraph
```
Comment (amitaibu) [inline Web/Controller/LandingPages.hs:59]: The error we get

image

File: Web/Controller/LandingPages.hs [hunk 3]
```diff
@@ -38,6 +38,44 @@ instance Controller LandingPagesController where
...
+                            where
+                                updateParagraph :: forall record.
+                                        ( HasField "id" record (Id record)
+                                        , SetField "weight" record Int
+                                        , CanUpdate record
+                                        , ?modelContext :: ModelContext
+                                        ) => record -> IO ()
+                                updateParagraph paragraph = do
+                                    let uuid = unpackId paragraph.id
+                                    let weight = elemIndex uuid uuids |> fromMaybe 0
```
Comment (amitaibu) [inline Web/Controller/LandingPages.hs:70]: image

File: Web/Controller/LandingPages.hs [hunk 4]
```diff
@@ -38,6 +38,44 @@ instance Controller LandingPagesController where
...
+
+                            where
+                                updateParagraph :: forall record.
+                                        ( HasField "id" record (Id record)
+                                        , SetField "weight" record Int
+                                        , CanUpdate record
+                                        , ?modelContext :: ModelContext
+                                        ) => record -> IO ()
+                                updateParagraph paragraph = do
+                                    let uuid = unpackId paragraph.id
```
Comment (amitaibu) [inline Web/Controller/LandingPages.hs:69]: image

File: Web/Controller/LandingPages.hs [hunk 5]
```diff
@@ -38,6 +38,46 @@ instance Controller LandingPagesController where
...
+                                updateParagraph :: forall record.
+                                        ( HasField "id" record (Id record)
+                                        , SetField "weight" record Int
+                                        , CanUpdate record
+                                        , Show (PrimaryKey (GetTableName record))
+                                        , ?modelContext :: ModelContext
+                                        ) => record -> IO ()
+                                updateParagraph paragraph = do
+                                    -- @todo: Why doesn't unpackId paragraph.id work here?
+                                    let uuid = show paragraph.id
```
Comment (amitaibu) [inline Web/Controller/LandingPages.hs:71]: If we try `unpackId` we get an error -- how to solve? Instead I've converted in the line below the Uuids to Text

image

Comment (mpscholten) [reply Web/Controller/LandingPages.hs:71]: It misses a constraint `PrimaryKey (GetTableName record) ~ UUID` (IHP also supports non UUID primary keys, so `unpackId` could potentially return something non-UUID, the constraint basically says that the primary key of that record needs to be an UUID).

Here's the full code that works on my clone of this project:

```haskell
                                updateParagraph :: forall record.
                                        ( HasField "id" record (Id record)
                                        , SetField "weight" record Int
                                        , CanUpdate record
                                        , Show (PrimaryKey (GetTableName record))
                                        , ?modelContext :: ModelContext
                                        , PrimaryKey (GetTableName record) ~ UUID
                                        ) => record -> IO ()
                                updateParagraph paragraph = do
                                    let uuid :: UUID = unpackId paragraph.id
                                    let weight = (elemIndex uuid uuids) |> fromMaybe 0

                                    paragraph
                                        |> set #weight weight
                                        |> updateRecord


                                    pure ()
```

Comment (amitaibu) [reply Web/Controller/LandingPages.hs:71]: Great, thank you. I'm new to this `~` - what does it mean in this context?

Comment (mpscholten) [reply Web/Controller/LandingPages.hs:71]: It's an equal constraint, so `a ~ b` means a should be equal to b at the type level

**Pull request merged**